delta/sdk/src/main/scala/ch/epfl/bluebrain/nexus/delta/sdk/auth/AuthTokenProvider.scala

```diff
@@ -0,0 +1,61 @@
+package ch.epfl.bluebrain.nexus.delta.sdk.auth
+
+import cats.effect.Clock
+import ch.epfl.bluebrain.nexus.delta.kernel.cache.KeyValueStore
+import ch.epfl.bluebrain.nexus.delta.kernel.utils.IOUtils
+import ch.epfl.bluebrain.nexus.delta.sdk.identities.model.AuthToken
+import monix.bio.UIO
+
+import java.time.{Duration, Instant}
+
+/**
+  * Provides an auth token for the service account, for use when comunicating with remote storage
+  */
+trait AuthTokenProvider {
+  def apply(): UIO[Option[AuthToken]]
+}
+
+object AuthTokenProvider {
+  def apply(auth: Option[AuthenticateAs], keycloakAuthService: KeycloakAuthService): AuthTokenProvider = {
+    auth match {
+      case Some(authAs) => new CachingKeycloakAuthTokenProvider(authAs, keycloakAuthService)
+      case None         => new AnonymousAuthTokenProvider
+    }
+  }
+  def anonymousForTest: AuthTokenProvider = new AnonymousAuthTokenProvider
+}
+
+private class AnonymousAuthTokenProvider extends AuthTokenProvider {
+  override def apply(): UIO[Option[AuthToken]] = UIO.pure(None)
+}
+
+private class CachingKeycloakAuthTokenProvider(identity: AuthenticateAs, service: KeycloakAuthService)(implicit
+    clock: Clock[UIO]
+) extends AuthTokenProvider {
+  private val cache = KeyValueStore.create[Unit, AccessTokenWithMetadata]()
```

Comment:
I wanted to use existing code, but I don't think this class is a great fit. We really want something where it deals with multiple threads accessing the value at once and not fetching X times

Reply:
It feels like there should be a scala library that can do this